modules/internal/DefaultAssociative.chpl

```diff
@@ -566,14 +566,15 @@ module DefaultAssociative {
 
     // ref version
     proc dsiAccess(idx : idxType, haveLock = false) ref {
-      const shouldLock = dom.parSafe && !haveLock;
+      const arrOwnsDom = dom._arrs.length == 1;
+      const shouldLock = dom.parSafe && !haveLock && arrOwnsDom;
```

Comment:
@benharsh stopped by to point out that even on master or in the current version of this branch, there's still a race even in my simple forall loop:

```chapel
forall key in collection do
  A[key] += 1;
```

Specifically, the `A[key]` expression gets a reference to a particular location, but then nothing prevents another simultaneous access from growing the domain and invalidating the reference before the += operation has a chance to update that reference.  So that also might argue for not doing any locking on any write-base dsiAccesses (which I think is what you were proposing a few messages above, and maybe in your last message as well).